Use global lock instead of per instance lock.

aazzolini opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30117 Use global lock instead of per instance lock.**

Multiple "trainers" could each create different instances of DistributedOptimizer, which means we can still have a race condition unless we do a trully global per worker lock.

Differential Revision: [D18603936](https://our.internmc.facebook.com/intern/diff/D18603936/)

aazzolini committed:
Use global lock instead of per instance lock.

Multiple "trainers" could each create different instances of DistributedOptimizer, which means we can still have a race condition unless we do a trully global per worker lock.

Differential Revision: [D18603936](https://our.internmc.facebook.com/intern/diff/D18603936/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/optim/optimizer.py:
```diff
@@ -6,17 +6,25 @@
+    # Ideally we would only need to share a lock for instances of
+    # _LocalOptimizer that deal with the same parameters. We are
+    # making a simplifying assumption here that if there is more
+    # than one instancde of _LocalOptimizer per worker, they will
```
instancde -> instance

mrshenli commented on torch/distributed/optim/optimizer.py:
```diff
@@ -6,17 +6,25 @@
+    # than one instancde of _LocalOptimizer per worker, they will
+    # be optimizing the same parameters (e.g. each data parallel
+    # trainer will create its own instance of _LocalOptimizer but
+    # they will all optimize the same parameters on each worker)
```
this is a strong assumption. Is there any workaround to drop this assumption and support per parameter lock?

aazzolini replied:
Workarounds would be too complex and not useful for most use-cases.

aazzolini committed:
Update on "Use global lock instead of per instance lock."


Multiple "trainers" could each create different instances of DistributedOptimizer, which means we can still have a race condition unless we do a trully global per worker lock.

Differential Revision: [D18603936](https://our.internmc.facebook.com/intern/diff/D18603936/)

[ghstack-poisoned]

aazzolini committed:
Update on "Use global lock instead of per instance lock."


Multiple "trainers" could each create different instances of DistributedOptimizer, which means we can still have a race condition unless we do a trully global per worker lock.

Differential Revision: [D18603936](https://our.internmc.facebook.com/intern/diff/D18603936/)

[ghstack-poisoned]

pritamdamania87 commented:
@aazzolini @mrshenli Can we land this PR? This is a pretty serious bug.

pietern commented:
The same change has been merged by @pritamdamania87 in #31404.
